### PR TITLE
fix nan mask in multicategorical

### LINKED ISSUE #301: NaN masking not working in multicategorical stype
This is because when we are doing slicing, we are creating a new tensor. So the inplace modification to the `feat.values` does not get reflected back on the original `feat` passed to the function.

### PR DESCRIPTION
**Pull request by yiweny**
Fixes #301
Will create a separate pr to improve the efficiency and remove implicit nans.

### CODE REVIEW

File: torch_frame/utils/compare.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            if key not in td2:
+                return False
+            if (not isinstance(td1, MultiEmbeddingTensor)
+                    and isinstance(td2, MultiEmbeddingTensor)) or (
+                        not isinstance(td1, MultiNestedTensor)
+                        and isinstance(td2, MultiNestedTensor)):
+                return False
+            if not _MultiTensor.allclose(td1[key], td2[key]):
+                return False
+        return True
```
Comment (weihua916) [inline torch_frame/utils/compare.py:41]: Do we need this (or are we using it)? `TensorFrame` already implements equality functionality, so I don't want to have a separate path for checking equality of tensor data here.

Comment (weihua916) [reply torch_frame/utils/compare.py:41]: see https://github.com/pyg-team/pytorch-frame/blob/master/torch_frame/data/tensor_frame.py#L199-L252

File: test/data/test_multi_embedding_tensor.py [hunk 2]
```diff
@@ -99,6 +99,39 @@ def test_from_tensor_list():
...
+    met_3 = met_with_nan.clone()
+    met_3.fillna_(torch.tensor([100, 101, 102]))
+    assert torch.all(
+        torch.eq(
+            met_3.values,
+            torch.tensor([
+                [100, 100, 100, 101, 101, 102],
+                [3, 4, 5, 8, 9, 11],
+            ])))
+    assert torch.all(torch.eq(met_3.offset, expected_offset))
```
Comment (weihua916) [inline test/data/test_multi_embedding_tensor.py:126]: Remove this since it's not used (and the behavior is confusing/unintuitive)

File: test/data/test_multi_nested_tensor.py [hunk 3]
```diff
@@ -81,6 +81,30 @@ def test_multi_nested_tensor_basics(device):
...
+    # Test fillna
+    multi_nested_tensor_2 = multi_nested_tensor_with_nan.clone()
+    multi_nested_tensor_2.fillna_(100)
+    assert not torch.all(multi_nested_tensor_2.values == -1).any()
+
+    # Test fillna with vector of shape (num_cols)
+    multi_nested_tensor_3 = multi_nested_tensor_with_nan.clone()
+    multi_nested_tensor_3.fillna_(torch.arange(100, 110))
+    assert torch.all(
+        torch.eq(multi_nested_tensor_3.offset, multi_nested_tensor.offset))
```
Comment (weihua916) [inline test/data/test_multi_nested_tensor.py:102]: I don't think we need to include this extra functionality, unless we use it in our main codebase.

File: test/data/test_multi_nested_tensor.py [hunk 4]
```diff
@@ -81,6 +81,30 @@ def test_multi_nested_tensor_basics(device):
     assert dense_multi_nested_tensor.shape == (multi_nested_tensor.shape[:-1] +
                                                (max_len, ))
 
+    multi_nested_tensor_with_nan = multi_nested_tensor.clone()
+    multi_nested_tensor_with_nan.values[0::2] = -1
```
Comment (weihua916) [inline test/data/test_multi_nested_tensor.py:85]: I am not sure what this line is doing. Can you clarify?

File: test/data/test_multi_embedding_tensor.py [hunk 5]
```diff
@@ -99,6 +99,39 @@ def test_from_tensor_list():
     assert torch.allclose(met.offset, expected_offset)
     assert_equal(tensor_list, met)
 
+    met_with_nan = met.clone()
+    met_with_nan.values[0::2] = -1
```
Comment (weihua916) [inline test/data/test_multi_embedding_tensor.py:103]: same here. Can you add a comment?

File: torch_frame/data/multi_embedding_tensor.py [hunk 6]
```diff
@@ -201,6 +201,18 @@ def _single_index_select(
             )
         assert False, "Should not reach here."
 
+    def fillna_col(
+        self,
+        index: int,
+        fill_value: Union[int, float, Tensor],
+    ) -> None:
```
Comment (weihua916) [inline torch_frame/data/multi_embedding_tensor.py:208]: Add doc-string

File: torch_frame/data/multi_nested_tensor.py [hunk 7]
```diff
@@ -275,6 +275,27 @@ def _single_index_select(self, index: int, dim: int) -> MultiNestedTensor:
         else:
             raise RuntimeError(f"Unsupported dim={dim} for index_select.")
 
+    def fillna_col(
+        self,
+        index: int,
+        fill_value: Union[int, float, Tensor],
+    ) -> None:
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:282]: Add doc-string.

File: torch_frame/data/multi_tensor.py [hunk 8]
```diff
@@ -325,6 +324,43 @@ def select(
...
+        """
+        if isinstance(fill_value, Tensor):
+            if fill_value.dim() != 1 or fill_value.numel() != self.num_cols:
+                raise ValueError(
+                    "fillna_ with Tensor of shape "
+                    f"{fill_value.shape} is not supported. "
+                    f"Please use a Tensor of shape ({self.num_cols}.)")
+            for col in range(self.num_cols):
+                self.fillna_col(col, fill_value[col])
+            return
```
Comment (weihua916) [inline torch_frame/data/multi_tensor.py:343]: From what I see, there is not much efficiency gain with this since we are doing for-loop over columns anyway. I'd just delete fill_value is Tensor case.

File: torch_frame/data/multi_tensor.py [hunk 9]
```diff
@@ -325,6 +324,43 @@ def select(
     def _single_index_select(self, index: int, dim: int) -> _MultiTensor:
         raise NotImplementedError
 
+    def fillna_(self, fill_value: Union[int, float, Tensor]):
+        """Fill the :obj:`MultiTensor` with fill_value.
+
+        Args:
+            index (int): A column index of the tensor to select.
```
Comment (weihua916) [inline torch_frame/data/multi_tensor.py:331]: delete `index` argument since we don't have one.

**Comment by weihua916** [comment]
Thanks! Two requests on my side

- Can we support the bare minimum in this PR: `fillna_col` for multi tensor, which is actually used. 
- Also, remove `eq` of `TensorData` since it's largely duplicated with that of `TensorFrame`

File: test/data/test_multi_nested_tensor.py (hunk 3 above)
Comment (yiweny) [reply test/data/test_multi_nested_tensor.py:102]: Sure.

File: test/data/test_multi_nested_tensor.py (hunk 4 above)
Comment (yiweny) [reply test/data/test_multi_nested_tensor.py:85]: Adding nan values to the tensor to test `fillna_col`.

File: test/data/test_multi_embedding_tensor.py (hunk 2 above)
Comment (yiweny) [reply test/data/test_multi_embedding_tensor.py:126]: Sure

File: test/data/test_multi_nested_tensor.py (hunk 4 above)
Comment (weihua916) [reply test/data/test_multi_nested_tensor.py:85]: Yes I understand that, I just did not understand why `0::2`, but maybe that's just arbitrary.

Comment (yiweny) [reply test/data/test_multi_nested_tensor.py:85]: Yeah it's just arbitrary

File: torch_frame/data/multi_tensor.py (hunk 8 above)
Comment (yiweny) [reply torch_frame/data/multi_tensor.py:343]: Sure.

**Comment by yiweny** [comment]
@weihua916 Hi. I removed `fillna_` and `eq` of `TensorData`

File: test/data/test_multi_embedding_tensor.py [hunk 10]
```diff
@@ -99,6 +99,21 @@ def test_from_tensor_list():
...
+
+    # Test fill_col
+    met_1 = met_with_nan.clone()
+    for col in range(met.num_cols):
+        met_1.fillna_col(col, 100)
+    assert not torch.all(met_1.values == -1).any()
+    expected_values = torch.tensor([
+        [100, 100, 100, 100, 100, 100],
+        [3, 4, 5, 8, 9, 11],
+    ])
```
Comment (weihua916) [inline test/data/test_multi_embedding_tensor.py:114]: I am having some hard time understanding test, especially the `expected_values`.  Can we add a test not by comparing the `values`, but compare based on elements, like `met_1[i, col]`?

Also, can we test the column-wise `fillna_col`, instead of filling all columns with 100?

File: test/data/test_multi_nested_tensor.py [hunk 11]
```diff
@@ -81,6 +81,18 @@ def test_multi_nested_tensor_basics(device):
...
+    # Set some values to be na
+    multi_nested_tensor_with_nan.values[0::2] = -1
+
+    # Test fill_col
+    multi_nested_tensor_1 = multi_nested_tensor_with_nan.clone()
+    for col in range(multi_nested_tensor_with_nan.num_cols):
+        multi_nested_tensor_1.fillna_col(col, 100)
+    assert not torch.all(multi_nested_tensor_1.values == -1).any()
+    assert torch.all(multi_nested_tensor_1.values[
+        multi_nested_tensor_with_nan.values.isnan()] == 100)
```
Comment (weihua916) [inline test/data/test_multi_nested_tensor.py:94]: Same here. Test for column-specific fillna

File: torch_frame/data/multi_nested_tensor.py [hunk 12]
```diff
@@ -275,6 +275,27 @@ def _single_index_select(self, index: int, dim: int) -> MultiNestedTensor:
         else:
             raise RuntimeError(f"Unsupported dim={dim} for index_select.")
 
+    def fillna_col(
+        self,
+        index: int,
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:280]: col_index

File: torch_frame/data/multi_nested_tensor.py [hunk 13]
```diff
@@ -275,6 +275,27 @@ def _single_index_select(self, index: int, dim: int) -> MultiNestedTensor:
...
+        )
+        diff = self.offset[start_idx + 1] - self.offset[start_idx]
+        batch, arange = _batched_arange(diff)
+        # Compute values
+        values = self.values[self.offset[start_idx][batch] + arange]
+        if self.values.is_floating_point():
+            values[torch.isnan(values)] = fill_value
+        else:
+            values[values == -1] = fill_value
+        self.values[self.offset[start_idx][batch] + arange] = values
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:297]: Is this necessary? the above in-place replacement is enough?

File: torch_frame/data/multi_tensor.py [hunk 14]
```diff
@@ -325,6 +324,21 @@ def select(
     def _single_index_select(self, index: int, dim: int) -> _MultiTensor:
         raise NotImplementedError
 
+    def fillna_col(
+        self,
+        index: int,
```
Comment (weihua916) [inline torch_frame/data/multi_tensor.py:329]: `col_index`

**Review by weihua916 — COMMENTED** [review]
Final round of comments.

File: torch_frame/data/multi_nested_tensor.py [hunk 15]
```diff
@@ -275,6 +275,27 @@ def _single_index_select(self, index: int, dim: int) -> MultiNestedTensor:
...
+        start_idx = torch.arange(
+            index,
+            self.num_rows * self.num_cols,
+            self.num_cols,
+            device=self.device,
+        )
+        diff = self.offset[start_idx + 1] - self.offset[start_idx]
+        batch, arange = _batched_arange(diff)
+        # Compute values
+        values = self.values[self.offset[start_idx][batch] + arange]
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:292]: rename to `values_col` as I understand this extracts `values` with the specified `col_index`.

File: test/data/test_multi_embedding_tensor.py (hunk 10 above)
Comment (yiweny) [reply test/data/test_multi_embedding_tensor.py:114]: sure

File: torch_frame/data/multi_nested_tensor.py (hunk 13 above)
Comment (yiweny) [reply torch_frame/data/multi_nested_tensor.py:297]: Yes. We are creating a new tensor when we select some indices from a tensor. But when we are slicing, we are only creating a view.

File: test/data/test_multi_nested_tensor.py (hunk 11 above)
Comment (yiweny) [reply test/data/test_multi_nested_tensor.py:94]: Cool. I changed the code. Now it fills different columns with different fill values and checks the correctness for each column.

File: torch_frame/data/multi_embedding_tensor.py [hunk 16]
```diff
@@ -201,6 +201,18 @@ def _single_index_select(
             )
         assert False, "Should not reach here."
 
+    def fillna_col(
+        self,
+        col_index: int,
+        fill_value: Union[int, float, Tensor],
+    ) -> None:
+        value_index = slice(self.offset[col_index], self.offset[col_index + 1])
+        values_col = self.values[:, value_index]
```
Comment (weihua916) [inline torch_frame/data/multi_embedding_tensor.py:210]: super nit: `value_index` --> `values_index`

File: torch_frame/data/multi_nested_tensor.py [hunk 17]
```diff
@@ -275,6 +275,27 @@ def _single_index_select(self, index: int, dim: int) -> MultiNestedTensor:
...
+        start_idx = torch.arange(
+            col_index,
+            self.num_rows * self.num_cols,
+            self.num_cols,
+            device=self.device,
+        )
+        diff = self.offset[start_idx + 1] - self.offset[start_idx]
+        batch, arange = _batched_arange(diff)
+        # Compute values
+        values_col = self.values[self.offset[start_idx][batch] + arange]
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:292]: nit: cache `self.offset[start_idx][batch] + arange` for efficiency and better readabiilty. 
```python
values_index = self.offset[start_idx][batch] + arange
values_col = self.values[values_index]
....
self.values[values_index] = values_col
```

File: torch_frame/nn/encoder/stype_encoder.py [hunk 18]
```diff
@@ -208,7 +207,20 @@ def na_forward(self, feat: TensorData) -> TensorData:
...
+        # Add better safeguard here to make sure nans are actually
+        # replaced, expecially when nans are represented as -1's. They are
+        # very hard to catch as they won't error out.
+        filled_values = feat
+        if isinstance(feat, _MultiTensor):
+            filled_values = feat.values
+        if filled_values.is_floating_point():
+            assert not torch.isnan(filled_values).any()
+        else:
+            assert not (filled_values == -1).any()
```
Comment (weihua916) [inline torch_frame/nn/encoder/stype_encoder.py:223]: Thanks for being extra careful! just curious, do you know when this can happen after this fix? or is it mainly for safeguarding?

File: test/data/test_multi_embedding_tensor.py [hunk 19]
```diff
@@ -79,6 +79,42 @@ def test_size():
...
+    assert not torch.all(met_with_nan.values == -1).any()
+    for col in range(met_with_nan.num_cols):
+        column = met_with_nan[:, col]
+        assert torch.all(column.values == col)
+
+    tensor_list = [
+        torch.tensor([[torch.nan, torch.nan, torch.nan],
+                      [torch.nan, torch.nan, torch.nan]]),
+        torch.tensor([[torch.nan, torch.nan], [torch.nan, torch.nan]]),
+    ]
```
Comment (weihua916) [inline test/data/test_multi_embedding_tensor.py:104]: can we add a bit of non-nan values in `tensor_list` (at least one for each colun)?

File: test/data/test_multi_nested_tensor.py [hunk 20]
```diff
@@ -32,6 +32,48 @@ def column_select(
...
+        assert torch.all(column.values == col)
+
+    tensor_list = [
+        [
+            torch.tensor([torch.nan, torch.nan]),
+            torch.tensor([torch.nan, torch.nan, torch.nan])
+        ],
+        [torch.tensor([torch.nan]),
+         torch.tensor([torch.nan, torch.nan])],
+    ]
```
Comment (weihua916) [inline test/data/test_multi_nested_tensor.py:62]: same here. Add a bit of non-nan values.

**Review by weihua916 — APPROVED** [review]
Thanks for addressing the comments! Looks much better now.

LGTM after addressing the current comments.

File: torch_frame/nn/encoder/stype_encoder.py (hunk 18 above)
Comment (weihua916) [reply torch_frame/nn/encoder/stype_encoder.py:223]: Also, I am wondering if this code can handle `dict[str, MultiNestedTensor]` (currently, we don't allow NaN since it's a tokenized results).

Comment (yiweny) [reply torch_frame/nn/encoder/stype_encoder.py:223]: It's mainly for safeguarding. Also I don't know if `dict[str, MultiNestedTensor]` will hit this code because we don't allow specifying `NAStrategy` for `text_tokenized`.

File: test/data/test_multi_embedding_tensor.py (hunk 19 above)
Comment (yiweny) [reply test/data/test_multi_embedding_tensor.py:104]: Sure.

File: torch_frame/nn/encoder/stype_encoder.py (hunk 18 above)
Comment (weihua916) [reply torch_frame/nn/encoder/stype_encoder.py:223]: > Also I don't know if dict[str, MultiNestedTensor] will hit this code because we don't allow specifying NAStrategy for text_tokenized.
Ok makes sense. Then no worries for now.

File: test/data/test_multi_embedding_tensor.py (hunk 19 above)
Comment (yiweny) [reply test/data/test_multi_embedding_tensor.py:104]: Added. Are you fine with non-nan values in each column equal the `fill_value`?

Comment (weihua916) [reply test/data/test_multi_embedding_tensor.py]: Ideally different.

Comment (yiweny) [reply test/data/test_multi_embedding_tensor.py]: Makes sense. Then it can test that the `fillna_col` is not filling the non_nan values. I have updated my code to reflect this.

**Comment by weihua916** [comment]
Looks great!
